### PR TITLE
[WIP] issue828 state syncing

### PR DESCRIPTION
**Pull request by ackratos**
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

For #828 

This PR is still working in progress, main functionality can "work" on our app but needs testing (multiple peer edge cases and performance) and complete thread safety. 

Just raise PR for community review/comments

### CODE REVIEW

**Comment by ackratos** [comment]
https://github.com/tendermint/tendermint/issues/828#issuecomment-459948058

**Comment by zmanian** [comment]
I just wanted to give you some feedback that a number of us have looked at your implementation and hopefully we can provide in depth technical feedback soon.

At a high level, one of the issues that's been important to us is that API for state synching must provide a mechanism of verifying chunks of data incrementally against the state root Bittorrent style.

This is one of our highest priority features in Cosmos over the next few months.

**Comment by ackratos** [comment]
> At a high level, one of the issues that's been important to us is that API for state synching must provide a mechanism of verifying chunks of data incrementally against the state root Bittorrent style.

Thanks for looking into this!
Yes, that's major missing part of this PR. The specification doesn't describe verify mechanism in detail, what I thought is it should be like an additional calculated check sum? i.e. concatenate chunks then check sum it?

**Comment by zmanian** [comment]
The basic approach we are thinking of is requiring the underlying data store to support RangeProofs like IAVL tree does and then provide a range proof for all the leaves in the chunk.

**Comment by ackratos** [comment]
> The basic approach we are thinking of is requiring the underlying data store to support RangeProofs like IAVL tree does and then provide a range proof for all the leaves in the chunk.

then we need a new abci api like `VerifyRecoveryChunk(chunk [][]byte) error`? Or write chunks as soon as we receive? (Currently this implementation write state chunks on when we received all chunks)

**Comment by zmanian** [comment]
Yeah I think VerifyRecoveryChunk makes sense.

If you only write chunks when all have been received, a malicious peer can trivially send you unlimited number of invalid chunks that you can't detect.

**Comment by ackratos** [comment]
Sorry for long time no update, but I was keep testing this against our application. It works (in preliminary stage, not tested in large scale) now on our environment, I have updated this PR to latest status so you can take another round of review.

And what's more, relying on iterate db on every state sync related abci calls will must be a heavy work. We want borrow the way ethereum parity warp-sync (https://wiki.parity.io/Warp-Sync) or BitTorrent organize snapshot (application db version) that load version - split it into chunk files with hash - have a  manifest know all parts' hash - peers download manifest + some files via p2p (at the same time, they can serve file parts to other nodes)

I haven't taken deep look into detail design/implementation, but want hear from your opinions on this proposal, feel free to review this PR and reply to our new idea:)

**Comment by ackratos** [comment]
> Sorry for long time no update, but I was keep testing this against our application. It works (in preliminary stage, not tested in large scale) now on our environment, I have updated this PR to latest status so you can take another round of review.
> 
> And what's more, relying on iterate db on state sync related abci (ReadChunks / LatestSnapshot) will must be a heavy work. We want borrow the way ethereum parity warp-sync (https://wiki.parity.io/Warp-Sync) or BitTorrent organize snapshot (application db version) that load version - split it into chunk files with hash - have a manifest know all parts' hash - peers download manifest + some files via p2p (at the same time, they can serve file parts to other nodes)
> 
> I haven't taken deep look into detail design/implementation, but want hear from your opinions on this proposal, feel free to review this PR and reply to our new idea:)

@melekes @ebuchman any idea on this?

File: types/params.go [hunk 1]
```diff
@@ -12,6 +12,12 @@ const (
 
 	// BlockPartSizeBytes is the size of one block part.
 	BlockPartSizeBytes = 65536 // 64kB
+
+	// MaxStateSizeBytes is the maximum permitted size of the blocks.
+	// TODO: reduce this by segmentation
+	MaxStateSizeBytes = 1048576000 // 1000MB
+
+	MonitorWindowInSeconds = 40
```
Comment (ebuchman) [inline types/params.go]: doesn't seem to below here. These are hard consensus params, but this is more like a reactor setting.

**Comment by ackratos** [comment]
> * making `LatestSnapshot` and `ReadSnapshotChunk` methods just be particular uses of `Query`. Query is already supposed to be used for reading from the current and historical state, so I think we should be able to reuse it for that here too. Do you think we really need the new methods, or can we just use Query?

abci query shares rw lock with other block execution apis (begin, end, commit, deliver, check tx) and this query is very heavy according to our experience. In case of chain have 500k user, there would be 1 million keys for acc substore of app state iterate all to get num of keys would cost several seconds...

we implemented cache for state sync in app level and only update the cache on syncable waypoint. The updating and query of cache shares a lock different with abci lock.

**Comment by ackratos** [comment]
> * removing `StartRecovery`. Do we really need it ?
> * change `WriteRecoveryChunk` to take a struct that also includes a `done bool` to indicate if this is the last chunk. then we don't need EndRecovery either!

Good idea, I can do this.

**Comment by ackratos** [comment]
> I would also think the state sync reactor should come in its own package. We can just have `blockchain/state` for now but we probably want to rename `blockchain` to `sync` and then have `sync/blockchain` and `sync/state` and move the existing `blockchain/store*` elsewhere. Just thinking aloud here and this shouldn't all happen in the same PR.

good idea. will do this.

**Comment by ackratos** [comment]
closed in favor of new implementation https://github.com/tendermint/tendermint/pull/3594 for new proposal https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit?usp=sharing
